dbms/src/Debug/MockStorage.cpp

```diff
@@ -0,0 +1,147 @@
+}
+
+bool MockStorage::tableExists(Int64 table_id)
+{
+    return table_schema.find(table_id) != table_schema.end();
+}
+
+ColumnsWithTypeAndName MockStorage::getColumns(Int64 table_id)
+{
+    if (tableExists(table_id))
+    {
+        return table_columns[table_id];
+    }
+    throw Exception(fmt::format("Failed to get columns by table_id '{}'", table_id));
+}
+
+MockColumnInfoVec MockStorage::getTableSchema(String name)
+{
+    if (tableExists(getTableId(name)))
+    {
+        return table_schema[getTableId(name)];
+    }
+    throw Exception(fmt::format("Failed to get table schema by table name '{}'", name));
+}
+
+/// for exchange receiver
+void MockStorage::addExchangeSchema(String & exchange_name, const MockColumnInfoVec & columnInfos)
+{
+    exchange_schemas[exchange_name] = columnInfos;
+}
+
+void MockStorage::addExchangeData(const String & exchange_name, ColumnsWithTypeAndName & columns)
+{
+    exchange_columns[exchange_name] = columns;
+}
+
+bool MockStorage::exchangeExists(const String & executor_id)
+{
+    return exchange_schemas.find(executor_id_to_name_map[executor_id]) != exchange_schemas.end();
+}
+
+bool MockStorage::exchangeExistsWithName(const String & name)
+{
+    return exchange_schemas.find(name) != exchange_schemas.end();
+}
+
+ColumnsWithTypeAndName MockStorage::getExchangeColumns(String & executor_id)
+{
+    if (exchangeExists(executor_id))
+    {
+        return exchange_columns[executor_id_to_name_map[executor_id]];
+    }
+    throw Exception(fmt::format("Failed to get exchange columns by executor_id '{}'", executor_id));
+}
+
+void MockStorage::addExchangeRelation(String & executor_id, String & exchange_name)
+{
+    executor_id_to_name_map[executor_id] = exchange_name;
+}
+
+MockColumnInfoVec MockStorage::getExchangeSchema(String exchange_name)
+{
+    if (exchangeExistsWithName(exchange_name))
+    {
+        return exchange_schemas[exchange_name];
+    }
+    throw Exception(fmt::format("Failed to get exchange schema by exchange name '{}'", exchange_name));
+}
+
+ColumnsWithTypeAndName MockStorage::getColumnsForMPPTableScan(Int64 table_id, Int64 partition_id, Int64 partition_num)
+{
+    if (tableExists(table_id))
+    {
+        auto columns_with_type_and_name = table_columns[table_id];
+        size_t rows = 0;
+        for (const auto & col : columns_with_type_and_name)
+        {
+            if (rows == 0)
+                rows = col.column->size();
+            assert(rows == col.column->size());
+        }
+
+        int start, per_rows, rows_left, cur_rows;
+        per_rows = rows / partition_num;
+        rows_left = rows - per_rows * partition_num;
+        if (partition_id >= rows_left)
+        {
+            start = (per_rows + 1) * rows_left + (partition_id - rows_left) * per_rows;
+            cur_rows = per_rows;
+        }
+        else
+        {
+            start = (per_rows + 1) * partition_id;
+            cur_rows = per_rows + 1;
+        }
```

Comment:
Now, data are split averagely to each mock mpp task. I think we can do this work through a function, so that we can control the amount of data each mock mpp task should receive, not averagely.

Reply:
> Now, data are split averagely to each mock mpp task. I think we can do this work through a function, so that we can control the amount of data each mock mpp task should receive, not averagely.

I mean that we can put this work into a single function so that we can adjust the behavior of sharding data more easily and it's more clear for the code readers.

Reply:
> > Now, data are split averagely to each mock mpp task. I think we can do this work through a function, so that we can control the amount of data each mock mpp task should receive, not averagely.
> 
> I mean that we can put this work into a single function so that we can adjust the behavior of sharding data more easily and it's more clear for the code readers.

Good idea, then we may mock the situation like data skew in an separate function